### PR TITLE
NPDA Groups and Permissions

### LINKED ISSUE #126: Permissions Discussion
[ discussion with @eatyourpeas ]

For clarification, we should ask the NPDA team if they are happy with the current set of permissions that have been assigned to roles (for reference, these are in the documentation site in the dev section in the Users page).

Some questions to pass on:

1. For the can opt out patient permission, is it okay for this to be assigned to lead clinician, npda audit team + superusers?
2. Should we use the CAN_LOCK_CHILD_PATIENT_DATA_FROM_EDITING permission? We shouldn't have to edit patients in previous cohorts, we'll only be able to edit patients in the current cohort and as such there's no need to lock patients like in E12.
3. Should the site administrator have access / be able to edit clinical data?

### LINKED ISSUE #134: Team discussion about permissions
[ discussion with @eatyourpeas @anchit-chandran @mbarton @AmaniKrayemRCPCH @cillian-rcpch ]

@AmaniKrayemRCPCH here is the schematic we have at the moment for NPDA permissions:

image

And here is some notes on what we discussed:

- Lead clinician = administrator (highest level of permission outside of RCPCH team). They can add users 
- Clinician users should not be able to change other user data or add any users. They will only be able to view other users in their organisation.

Feel free to expand / add anything on this below!

### LINKED ISSUE #146: is_staff chekcbox on NPDAUser form consideration
the is_staff button is to assign the privilege to a user to be able to access the django admin. However, considering that it is users of type Coordinator and the RCPCH audit team who will be utilising the NPDAUser form, and thus assigning these privileges, we should potentially remove/rename this box because it could lead to confusion. 

'Is the user I'm adding a member of staff at the hospital? Yes, so tick `is_staff`' - might be the kind of thought process giving our users accidental access to the django admin!

### PR DESCRIPTION
**Pull request by dc2007git**
### Overview

This PR integrates permissions and groups into NPDA. There are 4 user groups - reader, editor, coordinator, rcpch audit team (and rcpch children and family which has not been developed yet, but is there as a placeholder). Each of these groups has its own set of CRUD (create, read, update, delete) permissions explicitly defined in this PR, with regards to 4 model types - Patient, NPDAUser, Visit and Site. If a user attempts to access a view related to a permission that they do not have access to (say, the Update view which corresponds to the update permission), a 403 will be returned with the following message: 'You do not have the appropriate permissions to access this page/feature. Contact your Coordinator for assistance.'

Additionally, NPDA structures the NHS's organisations as PDUs. Each NPDAUser belongs to a PDU, as does each patient. As such, users should only be able to CRUD patients or visits related to their PDU - and not access any data that they want. This PR will integrate Django-based protection against both of these possibilities, by integrating PermissionRequired and PDU check mixins, to ensure that when a user goes to CRUD any patient, user, or visit, they can only do so if they are part of the same PDU (and of course have the permission to do that). 

That is the first layer of 'protection'. The second layer is much more to do with user simplicity. In summary - if a user does not have permission to delete a patient, for example, then the delete button will be hidden from them. This will reduce the number of users who have a 403 thrown and thus get confused - a great idea by @eatyourpeas .

Another key change is that when adding a user, a Coordinator can only grant another user the same or reduced permissions than what they have - so they cannot tick the is_staff or is_superuser button. Coordinators cannot add rcpch-audit-team members, for example.

And of course, a large refactor of code, removing references to clinician, lead clinician, administrator in favour of our new groups has been carried out.

## Documentation changes

Changed docs for new group refactor, permissions update

## Related issues

Closes #146 #134 #126 

Previous discussions on this PR are to be found on the old PR #129

### CODE REVIEW

File: documentation/docs/clinician-users/get-started.md [hunk 1]
```diff
@@ -7,13 +7,13 @@ reviewers: Dr Marcus Baw, Dr Simon Chapman
 
 ### Making a new account
 
-Lead Clinicians from Trusts/Health Boards will need to contact the NPDA team to create their account.
+Coordinators from Trusts/Health Boards will need to contact the NPDA team to create their account.
 
-Clinicians and Administrators can ask their Lead to create an account for them. Alternatively, they can contact the NPDA team, who can create their account with written permission from the relevant Trust/Health Board Lead Clinician.
+Clinicians and Administrators can ask their Lead to create an account for them. Alternatively, they can contact the NPDA team, who can create their account with written permission from the relevant Trust/Health Board Coordaintor.
```
Comment (mbarton) [inline documentation/docs/clinician-users/get-started.md]:
```suggestion
Clinicians and Administrators can ask their Lead to create an account for them. Alternatively, they can contact the NPDA team, who can create their account with written permission from the relevant Trust/Health Board Coordinator.
```

File: project/npda/general_functions/group_for_group.py [hunk 2]
```diff
@@ -26,11 +26,12 @@ def group_for_role(role_key):
     """
     Allocate Groups - the groups already have permissions allocated
     """
-    if role_key == AUDIT_CENTRE_LEAD_CLINICIAN:
-        group = Group.objects.get(name=TRUST_AUDIT_TEAM_FULL_ACCESS)
-    elif role_key == AUDIT_CENTRE_CLINICIAN:
+    # TODO change role keys
```
Comment (mbarton) [inline project/npda/general_functions/group_for_group.py]: Do we still need to change these?

File: project/npda/migrations/0002_alter_npdauser_role.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        ("npda", "0001_initial"),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name="npdauser",
+            name="role",
+            field=models.PositiveSmallIntegerField(
+                choices=[
+                    (1, "Coordinator"),
```
Comment (mbarton) [inline project/npda/migrations/0002_alter_npdauser_role.py]: Maybe reference the constants in user.py here so we have the string role descriptions in one place?

File: project/npda/views/npda_users.py [hunk 4]
```diff
@@ -288,19 +316,29 @@ def get_form_kwargs(self):
...
+        )
         context["title"] = "Edit NPDA User Details"
         context["button_title"] = "Edit NPDA User Details"
         context["form_method"] = "update"
         context["npda_user"] = NPDAUser.objects.get(pk=self.kwargs["pk"])
         return context
 
-    def is_valid(self):
-        return super(NPDAUserForm, self).is_valid()
+    # def is_valid(self):
```
Comment (mbarton) [inline project/npda/views/npda_users.py]: Minor: remove this?

File: project/npda/views/npda_users.py [hunk 5]
```diff
@@ -288,19 +316,27 @@ def get_form_kwargs(self):
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
+        context["show_rcpch_team_box"] = (
+            self.request.user.is_superuser
+            or self.request.user.is_rcpch_audit_team_member
```
Comment (dc2007git) [inline project/npda/views/npda_users.py:321]: @mbarton @eatyourpeas we have a checkbox for is_rcpch_audit_team_member and also a possible role allocation. Do we need both, is there an instance in which someone would select reader/editor/coordinator and still be in the audit team, ticking that box? And what would be the rationale, as opposed to simply selecting the actual audit team role from the dropdown?

**Review by eatyourpeas — APPROVED** [review]

**Pull request merged**